wallet/src/account/output_cache/mod.rs

```diff
@@ -732,56 +736,129 @@ impl OutputCache {
+                        frozen_token_id: None,
+                        confirmed_account_nonce: Some((cmd.into(), *nonce)),
+                    }),
+                    | AccountCommand::FreezeToken(token_id, _) => Some(Conflict {
+                        frozen_token_id: Some(*token_id),
+                        confirmed_account_nonce: Some((cmd.into(), *nonce)),
+                    }),
+                },
+            }
         });
 
-        let mut conflicting_txs = vec![];
-        if let Some(frozen_token_id) = frozen_token_id {
+        // Collect all conflicting txs
+        let mut conflicting_txs = BTreeSet::new();
+
+        if let Some(conflict) = conflict {
             for unconfirmed in self.unconfirmed_descendants.keys() {
                 let unconfirmed_tx = self.txs.get(unconfirmed).expect("must be present");
-                if self.uses_token(unconfirmed_tx, frozen_token_id) {
-                    let unconfirmed_tx = self.txs.get_mut(unconfirmed).expect("must be present");
-                    match unconfirmed_tx {
-                        WalletTx::Tx(ref mut tx) => {
-                            tx.set_state(TxState::Conflicted(block_id));
+
+                if let WalletTx::Tx(tx) = unconfirmed_tx {
+                    if let Some(frozen_token_id) = conflict.frozen_token_id {
+                        if self.uses_token(unconfirmed_tx, &frozen_token_id) {
+                            conflicting_txs.insert(tx.get_transaction().get_id());
                         }
-                        WalletTx::Block(_) => {}
-                    };
+                    }
 
-                    conflicting_txs.push(unconfirmed);
+                    if let Some((confirmed_account, confirmed_account_nonce)) =
+                        conflict.confirmed_account_nonce
+                    {
+                        if confirmed_tx.get_id() != tx.get_transaction().get_id()
+                            && uses_conflicting_nonce(
+                                unconfirmed_tx,
+                                confirmed_account,
+                                confirmed_account_nonce,
+                            )
+                        {
+                            conflicting_txs.insert(tx.get_transaction().get_id());
+                        }
+                    }
                 }
             }
         }
 
-        conflicting_txs
-            .into_iter()
-            .map(|tx_id| self.txs.get(tx_id).expect("must be present"))
-            .collect_vec()
+        // Remove all descendants of conflicting txs
+        let mut conflicting_txs_with_descendants = vec![];
+
+        for conflicting_tx in conflicting_txs {
+            let txs_to_rollback = self.remove_from_unconfirmed_descendants(conflicting_tx);
+
+            // Mark conflicting tx and its descendants as Conflicting and update OutputCache data accordingly
+            for tx_id in txs_to_rollback.iter().rev().copied() {
```

Comment:
There is a problem that currently I think cannot happen but can happen in the future when we add TXs from Mempool.

The problem is with the rollback_tx_data, as the TXs are iterated in random order by tx_id, in a situation like:
Having unconfirmed txs for a delegation or token:
A(nonce 1) -> B(2) -> C(3) -> D(4)
If for some reason the conflict happens in the middle of the chain e.g. nonce 2 (currently it should always happen at the start of the chain from confirmed blocks, unless I am mistaken)
So if it happens at 2, all 2, 3, and 4 will be inside conflicting_txs, but not sorted by tx_id. So the rollback can happen to finish at 3 or 4, which will be unable to find last_parent, the tx A as a still unconfirmed tx for this delegation or token.

Reply:
Thinking about it maybe you can have filter them out in n * log(d).
something like:
conflicting_descendents = BTreeSet::new();
for tx in conflicting_txs {
  conflicting_descendents += find_descendents(tx);
}
conflicting_txs = conflicting_txs.filter(|tx| conflicting_descendetns.contains(tx))

This should also fix the double Conflict -> Conflict state

Reply:
Am I getting it right that you suggest filter out conflicting tx to be able to rollback descendants first and then conflicting txs themself?

Reply:
txs_to_rollback  is already sorted with the parent first and all descendants after it, and the loop is in reverse.
only the conflicting_txs are not sorted, so if we filter them out before this loop, everything will be ok and there will be no duplicate txs and no Conflict -> Conflict state.

Reply:
So I checked the collection and I don't think filtering out parents is necessary as a result of `remove_from_unconfirmed_descendants`  function call doesn't contain any duplicates anymore.